### PR TITLE
Add binarySearch

### PR DESCRIPTION
**Pull request by kingdavidmartins**
### binarySearch

Use recursion to perform a logarithmic search similar to `.indexOf()` that finds the index of a value within an array. The two differences being 
1. Opperation only works with sorted arrays
2. Offers a major performance boost when compared liner search or `.indexOf()` 

```js
const binarySearch = (array, value, start = 0, end = array.length - 1) => {
  if (start > end) {
    return -1;
  }

  let middle = Math.floor((start + end) / 2);
  let value = array[middle];

  if (value > value) {
      return binarySearch(array, value, start, middle-1);
  }
  if (value < value) {
    return binarySearch(array, value, middle+1, end);
  }
  return middle;
}
```

```js
binarySearch([1, 4, 6, 7, 12, 13, 15, 18, 19, 20, 22, 24], 6); // 2
```

## What does your PR belong to?
- [ ] Website
- [x] Snippets
- [ ] General / Things regarding the repository (like CI Integration)

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (non-breaking improvement of a snippet)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## [Lodash Backlog](https://github.com/Chalarangelo/30-seconds-of-code/issues/100)

- [ ] I added the prefix [UPDATE: `method.md`] or [ADD: `method.md`]
- [ ] I have referenced the `method` to the lodash backlog.

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have checked that the changes are working properly
- [x] I have checked that there isn't any PR doing the same
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

**Comment by atomiks** [comment]
Looks nice, I modified it to align with our styling a bit more:

```js
const binarySearch = (arr, val, start = 0, end = arr.length - 1) => {
  if (start > end) return -1;
  const mid = Math.floor((start + end) / 2);
  const value = arr[mid];
  if (value > value) return binarySearch(arr, val, start, mid - 1);
  if (value < value) return binarySearch(arr, val, mid + 1, end);
  return mid;
}
```

Also, it would be really interesting to show a performance comparison between this and `indexOf` if you can include that in the example.

**Comment by kingdavidmartins** [comment]
Haha yea I noticed pushing update lol

**Comment by kingdavidmartins** [comment]
Now the following

### binarySearch

Use recursion to perform a logarithmic search similar to `.indexOf()` that finds the index of a value within an array. The two differences being 
1. Opperation only works with sorted arrays
2. Offers a major performance boost when compared liner search or `.indexOf()` 

```js
const binarySearch = (arr, val, start = 0, end = arr.length - 1) => {
  if (start > end) return -1;
  let mid = Math.floor((start + end) / 2);
  let target = arr[mid];
  if (target > val) return binarySearch(arr, val, start, mid-1);
  if (target < val) return binarySearch(arr, val, mid+1, end);
  return mid;
}
```

```js
binarySearch([1, 4, 6, 7, 12, 13, 15, 18, 19, 20, 22, 24], 6); // 2
```

File: snippets/binarySearch.md [hunk 1]
```diff
@@ -5,21 +5,13 @@ Use recursion to perform a logarithmic search similar to `.indexOf()` that finds
...
-      return binarySearch(array, value, start, middle-1);
-  }
-  if (value < value) {
-    return binarySearch(array, value, middle+1, end);
-  }
-  return middle;
+const binarySearch = (arr, val, start = 0, end = arr.length - 1) => {
+  if (start > end) return -1;
+  let mid = Math.floor((start + end) / 2);
+  let target = arr[mid];
```
Comment (atomiks) [inline snippets/binarySearch.md]: You can use `const` here, no?

**Comment by skatcat31** [comment]
I think we may want to hold off on merging this until we do the great re-categorization to put this under academic( limited application to type and pre sorted arrays )

**Comment by kingdavidmartins** [comment]
Now the following 

### binarySearch

Use recursion to perform a logarithmic search similar to `.indexOf()` that finds the index of a value within an array. The two differences being 
1. Operation only works with sorted arrays
2. Offers a major performance boost when compared to a linear search or `.indexOf()` 

```js
const binarySearch = (arr, val, start = 0, end = arr.length - 1) => {
  if (start > end) return -1;
  const mid = Math.floor((start + end) / 2);
  const target = arr[mid];
  if (target > val) return binarySearch(arr, val, start, mid-1);
  if (target < val) return binarySearch(arr, val, mid+1, end);
  return mid;
}
```

```js
binarySearch([1, 4, 6, 7, 12, 13, 15, 18, 19, 20, 22, 24], 6); // 2
binarySearch([1, 4, 6, 7, 12, 13, 15, 18, 19, 20, 22, 24], 21); // -1
binarySearch([1, 4, 6, 7, 12, 13, 15, 18, 19, 20, 22, 24], 20); // 9
```

**Comment by atomiks** [comment]
Here's a contrived example to showcase perf difference

```js
const val = 10000000; // 10 million
const testArr = Array(val + 1).fill().map((_, i) => i);

console.time('perf');
testArr.indexOf(val);
console.timeEnd('perf'); // 13.6ms

console.time('perf');
binarySearch(testArr, val);
console.timeEnd('perf'); // 0.02ms
```

**Comment by kingdavidmartins** [comment]
@skatcat31 I don't believe this falls under the academic criteria. This is something that gets used in production a great deal. Although numbers need to be "sorted" or "in ascending order". Let's Imagine a DB and/or collection that holds user ID's or Items ID's or a Dictionary of sorts. As you can all imagine when user's a generated with unique ID's they are done so being linear/consecutive/ascending in nature

Example:

```js
binarySearch(['apple', 'banana', 'candy', 'cool', 'dad', 'doom', 'energy', 'happy', 'watermelon', 'zoo'], 'watermelon')
```

So it would make sense if operating at scale *Millions/billions of users* to search using this method

**Comment by skatcat31** [comment]
@kingdavidmartins agreed these traversables get used a great deal in real world applications, most notably being databases.

I am talking about in categorization of our snippets though. Considering a Binary Search algorithm can be implemented regardless of datatype is why I think it should go under the "Academic" category since it might not actually be an Array, and it's purpose is to more teach an algorithm than a JS feature.

Binary Search Trees for example are Data Structure designed AROUND the binary search algorith to have a self sorting structure:

```js
const BSTBS = (bst, val) => {
  return bst.value == val ? bst : bst.left && bst.value < val ? BSTBS(bst.left, val) : bst.right && bst.value > val ? BSTBS(bst.right, val) : undefined
}
```

File: snippets/binarySearch.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+2. Offers a major performance boost when compared to a linear search or `.indexOf()` 
+
+```js
+const binarySearch = (arr, val, start = 0, end = arr.length - 1) => {
+  if (start > end) return -1;
+  const mid = Math.floor((start + end) / 2);
+  const target = arr[mid];
+  if (target > val) return binarySearch(arr, val, start, mid-1);
+  if (target < val) return binarySearch(arr, val, mid+1, end);
+  return mid;
```
Comment (Chalarangelo) [inline snippets/binarySearch.md]: Lines 12-14 could be one long ternary operator statement. Also, is there a need to assign `target` in the first place? You are not using it anywhere except these conditionals. I would argue the same about `mid`, but that would make the code slightly less readable and you would have to calculate it like 5 times, which is a performance hit on a performance-centric snippet, so I think that should remain as-is.

File: snippets/binarySearch.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
+### binarySearch
+
+Use recursion to perform a logarithmic search similar to `.indexOf()` that finds the index of a value within an array. The two differences being
```
Comment (Chalarangelo) [inline snippets/binarySearch.md]: `Array.indexOf()`

File: snippets/binarySearch.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
+### binarySearch
+
+Use recursion to perform a logarithmic search similar to `.indexOf()` that finds the index of a value within an array. The two differences being 
+1. Operation only works with sorted arrays
+2. Offers a major performance boost when compared to a linear search or `.indexOf()` 
+
```
Comment (Chalarangelo) [inline snippets/binarySearch.md]: You need to add an explanation on how the snippet works. Remember to say at the very top `Uses recursion.` and then explain the rest of the code. It's pretty simple, but it needs to be explained for readers to grasp.

File: snippets/binarySearch.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  if (target > val) return binarySearch(arr, val, start, mid-1);
+  if (target < val) return binarySearch(arr, val, mid+1, end);
+  return mid;
+}
+```
+
+```js
+binarySearch([1, 4, 6, 7, 12, 13, 15, 18, 19, 20, 22, 24], 6); // 2
+binarySearch([1, 4, 6, 7, 12, 13, 15, 18, 19, 20, 22, 24], 21); // -1
+binarySearch([1, 4, 6, 7, 12, 13, 15, 18, 19, 20, 22, 24], 20); // 9
```
Comment (Chalarangelo) [inline snippets/binarySearch.md]: No need for the third example, it doesn't showcase any extra use-cases that the previous two don't cover.

File: snippets/binarySearch.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
+### binarySearch
+
+Use recursion to perform a logarithmic search similar to `.indexOf()` that finds the index of a value within an array. The two differences being 
+1. Operation only works with sorted arrays
```
Comment (Chalarangelo) [inline snippets/binarySearch.md]: Only arrays sorted in ascending order, if I understand this correctly.

**Comment by Chalarangelo** [comment]
@skatcat31 Very good point about tagging as academic (`algorithms` is arguably a better name), but I think if `quicksort` is going to be archived, this snippet will probably meet the same fate (although this has more practical use-cases).

**Comment by Pl4gue** [comment]
I also think we should be careful with including algorithms. Our main suppose is bundling useful operations into small JS snippets, but there are more than enough algorithm libraries out there... @Chalarangelo

**Comment by Chalarangelo** [comment]
@Pl4gue I suppose that keeping algorithmic snippets in the archive is a win-win, as we do not have to showcase them, but they are still available and the effort people put into them doesn't go to waste.

**Comment by kingdavidmartins** [comment]
@Chalarangelo I'm going to just quickly do a new branch from master & do a PR. After doing a reset since you started doing multi tagging had to resolve conflicts which  README ended up reflecting changes to the README still

**Pull request closed without merging**